packages/hebao_v1/contracts/modules/base/BaseModule.sol

```diff
@@ -25,20 +25,14 @@ abstract contract BaseModule is Module
     event Activated   (address wallet);
     event Deactivated (address wallet);
 
-    struct ControllerCache
-    {
-        PriceOracle      priceOracle;
-        address          collectTo;
-    }
-
-    ControllerCache public controllerCache;
-
     ModuleRegistry public immutable moduleRegistry;
     SecurityStore  public immutable securityStore;
     WhitelistStore public immutable whitelistStore;
     QuotaStore     public immutable quotaStore;
     HashStore      public immutable hashStore;
     address        public immutable walletFactory;
+    PriceOracle    public immutable priceOracle;
```

Comment:
If the priceOracle is immutable, we have to make sure it's fully configurable after deployment, even the EXPIRY_PERIOD should be updatable.

To save more gas, I'd suggest we merge PriceCacheStore and Uniswap2PriceOracle into one single priceOracle contract, and still keep priceOracle cached in each module.

Reply:
> If the priceOracle is immutable, we have to make sure it's fully configurable after deployment, even the EXPIRY_PERIOD should be updatable.

Do we care a lot about `EXPIRY_PERIOD`? It doesn't seem that important, anyone is able to update the prices at any time with the oracle, or we can update the prices manually. It doesn't seem to be an important value. To me it's more like when in the worst case maybe a too high a price is set for a long time or something, but even then that's not too bad and we can fix that by overriding the price if we need to.

Reply:
> To save more gas, I'd suggest we merge PriceCacheStore and Uniswap2PriceOracle into one single priceOracle contract, and still keep priceOracle cached in each module.

`tokenValue()` doesn't read in the active price oracle so no gas is saved on the frequent path by merging them (only read in when `updateTokenPrice` is called, which won't be that often, I assume at most called once a day for some tokens) , so I don't think we should merge them so we are still able to update the actual price oracle when needed.

Reply:
> To save more gas, I'd suggest we merge PriceCacheStore and Uniswap2PriceOracle into one single priceOracle contract, and still keep priceOracle cached in each module.

Merging `QuotaStore` and `PriceCacheStore` would save a bit of gas however.

Reply:
Fine with me to keep the changes you made to priceOracle (immutable).